### PR TITLE
Assignment

### PR DESCRIPTION
**Pull request by cresparza**
Some tradeoffs of the PR:
The large output data provided and the output data generated by this service were the same, but had a different order. This caused the test to break. You can compare the JSON outputs and it has 39 differences by just getting some properties in different position. So an output with the same order was set. To pass the test

The most important trade off is that this service is only dealing with the item_id being in the last position and not mixed (first or last)

This could be fixed by moving the record elements to n-1 and the first position being moved to an aux variable, which will be then set in the last column.

### CODE REVIEW

**Review by nahojer — COMMENTED** [review]
Good work! Left some conversation points and suggestions.

General notes:
* Would love to see some tests written by you (even though the assignment already came  with e2e tests.)

General questions:
* Is the code production ready yet? If not, what is left to make it production ready?
* What's your perspective on flat file/package structure vs more atomic packages?

**Comment by cresparza** [comment]
> Good work! Left some conversation points and suggestions.
> 
> General notes:
> 
> * Would love to see some tests written by you (even though the assignment already came  with e2e tests.)
> 
> General questions:
> 
> * Is the code production ready yet? If not, what is left to make it production ready?
> * What's your perspective on flat file/package structure vs more atomic packages?

Hey Johan, thanks for the review, I am still relatively new working with Golang and its best practices. The comments are really helpful.

In its current state the project would not be production ready as it is, some very important things would need to be addressed first would be:

- To add unit test for the middleware and service function (minimum)
- To add slogs
- Validate the content-type and restrict the http methods to only GET
- Improve the error handling. 
- Fix for example the err comparison by using errors.Is functionality
- Fix the http response in case of error. Only returning the text corresponding 404, 501, etc and leaving the slogs for something more specific and helpful for devs

Then if we consider that fixed, I think as a release of a module would imply to release a version, I would prefer this version to follow the naming conventions too, so the renaming of files and functions would be important too, even if those changes are non blocking

About the middleware validation. My initial idea was to have basic validations of the csv structure, before processing it and make more complex operations, if it is an invalid, empty or the body has a different content type, then it should not reach the service. 

An idea that I left out was to rearrange the csv (if required) in another middleware function after validation  but once I started checking the possible solutions for this, I noticed that I needed basically 2 csv objects, the one being consumed during reading and the destination csv. I thought this might not be so optimal after all and it could happen during validation, but again I did not see any real improvement over making directly in the service as it is currently implemented. But I feel curious about exploring more optimal solutions.

File: internal/middlewares/csvMiddleware.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+		if index == 0 {
+			headerLength = len(record)
+			rearrange = utils.IsRearrangeRequired(record[index])
+			index++
+		}
+
+		// Rearrange item_id from first position to last position
+		if rearrange {
+			record = utils.RearrangeCSVRecord(record)
+		}
```
Comment (jaanusjaeger) [inline internal/middlewares/csvMiddleware.go:87]: Here it is assumed that if rearrange is not required (`item_id` is not the first column), the header is correct, but in fact `item_id` can also be missing and in that case it's assumed to be the last column.

**Review by jaanusjaeger — COMMENTED** [review]
Thank you for taking the time to complete the assignment.

It was nice to see `Dockerfile` and health check endpoints. Also that the server
logged it's activity, but I agree with Johan's comment that using a logging package
would be preferred for e.g configuring (turning off) logging on global level.

I added some comments into the code. As a general comment, it was good that you brought out the tradeoffs in the PR description, but how would you support any column order, where `item_id` can appear on any position?

File: internal/server/server.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+)
+
+type Server struct {
+	Router *mux.Router
+	Config *config.Config
+}
+
+func NewServer() *Server {
+	// Load config from file
+	serverConf := config.LoadConfig()
```
[the remark this answers is not in the document]
Comment (cresparza) [reply internal/server/server.go:19]: Thanks Jaanus I have not thought about it

File: internal/middlewares/csvMiddleware.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	"github.com/cresparza/csv-to-json/internal/utils"
+)
+
+var ErrorDepth = errors.New("erroneous non-N/A at higher depth")
+
+// ValidateCSV middleware validating a csv from the request body
+func ValidateCSV(next http.HandlerFunc) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Check if the request body is empty
+		if r.Body == nil {
```
[the remark this answers is not in the document]
Comment (cresparza) [reply internal/middlewares/csvMiddleware.go:20]: so maybe to check the r.ContentLength?

File: internal/middlewares/csvMiddleware.go (hunk 1 above)
Comment (cresparza) [reply internal/middlewares/csvMiddleware.go:87]: Thats true, I did not consider more scenarios. It is a nice catch. I guess then, we can discuss if the scenario with item_id not being present is valid or invalid. Then to add that scenario to the validation.

For the scenario with item_id in any position, then, the solution that comes to my mind quickly would be to slice the record array from position 0 to item_id index (without including it), then from item_id index +  to the end. Append slice 2 to slice 1 and  finally append the item_id

File: internal/service/csvService.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+// CreateJSONOutput take the request body and returns a marshalled JSON
+func CreateJSONOutput(requestBody io.ReadCloser) ([]byte, error) {
+	// Takes the request body (csv) and cast it to a string
+	bodyStr, err := utils.PayloadToString(requestBody)
+	if err != nil {
+		return nil, err
+	}
+
+	csvReader := csv.NewReader(strings.NewReader(bodyStr))
```
[the remark this answers is not in the document]
Comment (cresparza) [reply internal/service/csvService.go:26]: The problem that I faced by doing that, was that the string does not implement the io.Reader
